### PR TITLE
Samples

### PR DESCRIPTION
**Pull request by sokrahta**
+ update /suggest endpoint to include sample top X
 + gives back suggestion style results, but with no seeded query text
 + sorted by most frequent document count
+ unified suggest and sample response syntax

[Review on Reviewable](https://reviewable.io/reviews/socrata/spandex/30)

### CODE REVIEW

File: spandex-common/src/main/scala/com/socrata/spandex/common/client/SpandexElasticSearchClient.scala [hunk 1]
```diff
@@ -46,6 +46,8 @@ class SpandexElasticSearchClient(config: ElasticSearchConfig) extends ElasticSea
     boolQuery().must(termQuery(SpandexFields.DatasetId, datasetId))
                .must(termQuery(SpandexFields.CopyNumber, copyNumber))
                .must(termQuery(SpandexFields.ColumnId, columnId))
+  protected def byColumnCompositeId(column: Columnmap): QueryBuilder =
```
Comment (urmilan) [inline spandex-common/src/main/scala/com/socrata/spandex/common/client/SpandexElasticSearchClient.scala]: Is `Columnmap` different to `ColumnMap`?

File: spandex-common/src/main/scala/com/socrata/spandex/common/client/SpandexElasticSearchClient.scala [hunk 2]
```diff
@@ -275,11 +277,19 @@ class SpandexElasticSearchClient(config: ElasticSearchConfig) extends ElasticSea
...
-      .prepareSuggest(config.index)
+    val response = client.prepareSuggest(config.index)
       .addSuggestion(suggestion)
       .execute().actionGet()
 
     response.getSuggest
   }
+
+  // TODO sort frequency descending
+  def getSamples(column: ColumnMap, size: Int): SearchResults[FieldValue] = {
```
Comment (urmilan) [inline spandex-common/src/main/scala/com/socrata/spandex/common/client/SpandexElasticSearchClient.scala]: Did you mean to use the `size` param in the query?

File: spandex-common/src/main/scala/com/socrata/spandex/common/client/SpandexElasticSearchClient.scala (hunk 1 above)
Comment (sokrahta) [inline spandex-common/src/main/scala/com/socrata/spandex/common/client/SpandexElasticSearchClient.scala]: typo, 8cd5868

File: spandex-common/src/test/scala/com/socrata/spandex/common/client/SpandexElasticSearchClientSpec.scala [hunk 3]
```diff
@@ -246,4 +245,17 @@ class SpandexElasticSearchClientSpec extends FunSuiteLike with Matchers with Bef
     suggestionsUpper.toString should include(food.value)
     suggestionsUpper.toString should include(fool.value)
   }
+
+  test("samples") {
```
Comment (urmilan) [inline spandex-common/src/test/scala/com/socrata/spandex/common/client/SpandexElasticSearchClientSpec.scala]: Once we use the `size` param as mentioned above, let's add a test to ensure that if we request more than the average page size (10), and that many actually column maps exist, we return that many results. I wrote a similar test [here](https://github.com/socrata/spandex/pull/24/files#diff-888879e4382c1d761e64374ad1496e7aR131).

File: spandex-common/src/main/scala/com/socrata/spandex/common/client/SpandexElasticSearchClient.scala (hunk 2 above)
Comment (sokrahta) [inline spandex-common/src/main/scala/com/socrata/spandex/common/client/SpandexElasticSearchClient.scala]: yes, 4d65572

File: spandex-http/src/main/scala/com/socrata/spandex/http/SpandexServlet.scala [hunk 4]
```diff
@@ -32,29 +36,53 @@ class SpandexServlet(conf: SpandexConfig,
...
   }
 
-  get ("/suggest/:datasetId/:copyNum/:userColumnId/:text/?") {
+  private[this] val paramDatasetId = "datasetId"
+  private[this] val paramCopyNum = "copyNum"
+  private[this] val copyNumMustBeNumeric = "Copy number must be numeric"
+  private[this] val paramUserColumnId = "userColumnId"
+  private[this] val paramText = "text"
+  private[this] val paramFuzz = "fuzz"
+  private[this] val paramFize = "size"
```
Comment (urmilan) [inline spandex-http/src/main/scala/com/socrata/spandex/http/SpandexServlet.scala]: Typo in variable name :)

File: spandex-http/src/main/scala/com/socrata/spandex/http/SpandexServlet.scala [hunk 5]
```diff
@@ -32,29 +36,53 @@ class SpandexServlet(conf: SpandexConfig,
...
+    val datasetId = params.get(paramDatasetId).get
+    val copyNum = Try(params.get(paramCopyNum).get.toLong)
+      .getOrElse(halt(HttpStatus.SC_BAD_REQUEST, copyNumMustBeNumeric))
+    val userColumnId = params.get(paramUserColumnId).get
+    val size = params.get(paramFize).headOption.map{_.toInt}.getOrElse(conf.suggestSize)
+
+    val column = columnMap(datasetId, copyNum, userColumnId)
+
+    logger.info(s"GET /sample ${column.docId} / $paramFize:$size")
+    client.getSamples(column, size).thisPage.map { src =>
```
Comment (urmilan) [inline spandex-http/src/main/scala/com/socrata/spandex/http/SpandexServlet.scala]: As we discussed yesterday, let's make this a case class with JsonCodec that's used by both the /sample and /suggest endpoints.

File: spandex-http/src/main/scala/com/socrata/spandex/http/SpandexServlet.scala (hunk 4 above)
Comment (sokrahta) [inline spandex-http/src/main/scala/com/socrata/spandex/http/SpandexServlet.scala]: oh my gosh I can't focus; 6dc303c

File: spandex-http/src/main/scala/com/socrata/spandex/http/SpandexServlet.scala (hunk 5 above)
Comment (sokrahta) [inline spandex-http/src/main/scala/com/socrata/spandex/http/SpandexServlet.scala]: yup still in progress.

File: spandex-http/src/test/scala/com/socrata/spandex/http/SpandexServletSpec.scala [hunk 6]
```diff
@@ -76,7 +77,7 @@ class SpandexServletSpec extends ScalatraSuite with FunSuiteLike with TestESData
       status should equal(HttpStatus.SC_OK)
       val contentType: String = header.getOrElse(ContentTypeHeader, "")
       contentType should include(ContentTypeJson)
-      body should include("\"options\"")
```
Comment (urmilan) [inline spandex-http/src/test/scala/com/socrata/spandex/http/SpandexServletSpec.scala]: Once we define a response case class, we can replace all these strings with an object eg.
```scala
val expected = Response(Seq(Suggestion("data column 3 row 1"),Suggestion("data column 3 row 2")))
body should be (JsonUtil.renderJson(expected))
```

File: spandex-http/src/main/scala/com/socrata/spandex/http/SpandexServlet.scala [hunk 7]
```diff
@@ -32,29 +36,53 @@ class SpandexServlet(conf: SpandexConfig,
...
+    val size = params.get(paramSize).headOption.map{_.toInt}.getOrElse(conf.suggestSize)
+
+    val column = columnMap(datasetId, copyNum, userColumnId)
+
+    logger.info(s"GET /sample ${column.docId} / $paramSize:$size")
+    client.getSamples(column, size).thisPage.map { src =>
+      """{
+        |"text" : "%s",
+        |"score" : "%s"
+        |}""".stripMargin.format(src.value, src)
```
Comment (urmilan) [inline spandex-http/src/main/scala/com/socrata/spandex/http/SpandexServlet.scala]: Let's just leave out the score field if there's nothing relevant to return in this case.

File: spandex-common/src/test/scala/com/socrata/spandex/common/client/SpandexElasticSearchClientSpec.scala (hunk 3 above)
Comment (sokrahta) [inline spandex-common/src/test/scala/com/socrata/spandex/common/client/SpandexElasticSearchClientSpec.scala]: yup true; 9495045

File: spandex-http/src/test/scala/com/socrata/spandex/http/SpandexServletSpec.scala (hunk 6 above)
Comment (sokrahta) [inline spandex-http/src/test/scala/com/socrata/spandex/http/SpandexServletSpec.scala]: we can definitely put that on the backlog :smile:

File: spandex-common/src/main/resources/reference.conf [hunk 8]
```diff
@@ -65,11 +64,11 @@ com.socrata.spandex {
                     "type" : "category",
                     "path" : "composite_id"
                   }
-                }
+                },
+                "fields": { "raw": { "type": "string", "index": "not_analyzed" } }
```
Comment (urmilan) [inline spandex-common/src/main/resources/reference.conf]: This needs to get indexed somehow. I think it should work if you just tweak the JSON codec for FieldValue.

Comment (sokrahta) [inline spandex-common/src/main/resources/reference.conf]: removing the extra field.

File: spandex-http/src/main/scala/com/socrata/spandex/http/SpandexServlet.scala (hunk 7 above)
Comment (sokrahta) [inline spandex-http/src/main/scala/com/socrata/spandex/http/SpandexServlet.scala]: the score field is an `Option` and rojoma json omits if appropriates.

File: spandex-common/src/test/scala/com/socrata/spandex/common/client/SpandexElasticSearchClientSpec.scala [hunk 9]
```diff
@@ -258,23 +257,76 @@ class SpandexElasticSearchClientSpec extends FunSuiteLike with Matchers with Bef
...
+
+    samples.totalHits should be(5)
+    samples.aggs should contain(BucketCount(makeRowData(1, 1), 1))
+    samples.aggs should contain(BucketCount(makeRowData(1, 2), 1))
+    samples.aggs should contain(BucketCount(makeRowData(1, 3), 1))
+    samples.aggs should contain(BucketCount(makeRowData(1, 4), 1))
+    samples.aggs should contain(BucketCount(makeRowData(1, 5), 1))
+  }
+
+  ignore("lots of samples") {
```
Comment (urmilan) [inline spandex-common/src/test/scala/com/socrata/spandex/common/client/SpandexElasticSearchClientSpec.scala]: We shouldnt ignore this test, just modify it to not check the counts and ordering.

Comment (sokrahta) [inline spandex-common/src/test/scala/com/socrata/spandex/common/client/SpandexElasticSearchClientSpec.scala]: well actually, with the ES mappings reverted I don't think it's possible to get these donut coconuts at all.

**Comment by urmilan** [comment]
Ok, farewell dear donut coconuts.... :+1:

**Pull request merged**